Content.Client/GameObjects/EntitySystems/VerbSystem.cs

```diff
@@ -116,13 +119,39 @@ private bool OnOpenContextMenu(in PointerInputCmdHandler.PointerInputCmdArgs arg
             }
 
             var mapCoordinates = args.Coordinates.ToMap(_mapManager);
+            var playerEntity = _playerManager.LocalPlayer?.ControlledEntity;
+            if (playerEntity == null)
+            {
+                return false;
+            }
+            
             var entities = _entityManager.GetEntitiesIntersecting(mapCoordinates.MapId,
                 Box2.CenteredAround(mapCoordinates.Position, (0.5f, 0.5f))).ToList();
-
+            
             if (entities.Count == 0)
             {
                 return false;
             }
+            
+            // Check if we have LOS to the clicked-location, otherwise no popup.
+            var vectorDiff = playerEntity.Transform.MapPosition.Position - mapCoordinates.Position;
+            var direction = vectorDiff.Normalized;
+            var distance = vectorDiff.Length;
+            Func<IEntity, bool> ignored = entity => entities.Contains(entity) || 
+                                                    entity == playerEntity || 
+                                                    !entity.TryGetComponent(out OccluderComponent occluder) ||
+                                                    !occluder.Enabled;
+            
+            var ray = _physicsManager
+                .IntersectRayWithPredicate(mapCoordinates.MapId, 
+                    new CollisionRay(mapCoordinates.Position, direction, (int) CollisionGroup.Opaque), 
+                    distance, 
+                    ignored);
+
+            if (ray.Any())
+            {
+                return false;
+            }
```

Comment:
I feel like this check might come in handy in other places, so maybe you could add it somewhere as a helper for ease-of-use?

Reply:
Also is there an equivalent check server-side?

Reply:
There's already InRangeUnobstructed although the reason i didn't use it is because it doesn't return on first hit which seemed wasteful. I could expose the arg and just use InRangeUnobstructed.

As for server-side I was 50/50 on adding it because the entity's already on the client-side via PVS though I guess it could help people cheating to open the context menu so I'll add it in.

Reply:
I've had a look around; the server only sends back the verb context and not the initial opening (the client only sends a request when they do the 2nd right-click on an entity's context).

The difficulty with this is the server currently has no idea of what entities the player has open in their context menu at the moment so I'm not sure how to handle that for the ray. The client might have to send through the click point to know what entities can be ignored in range... any ideas?

Reply:
Okay I've come up with A solution, it may not necessarily be the BEST solution but the server will do a rough check to see if the verb request is valid.